From 8fddd571ef6d38115d1bc868722ebb0c22b93c37 Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Fri, 9 Oct 2026 00:07:59 +0200
Subject: [PATCH] Bindings: layout of the material data managers

C: mgis_bv_FieldLayout, mgis_bv_material_data_manager_initializer_set_layout,
mgis_bv_material_state_manager_get_layout. Fortran: POINT_BY_POINT and
COMPONENT_BY_COMPONENT, material_data_manager_initializer_set_layout,
material_state_manager_get_layout; the functions returning the arrays of a
material state manager as two-dimensional arrays (one column per point)
report an error for the component by component layout, which they can't
describe. Python: FieldLayout, layout members, numpy arrays with the
strides of the layout (same indices for both layouts), constructor of the
initializer of a material data manager, which could not be built. Tests in
the three languages.
---
 .../MGIS/Behaviour/MaterialDataManager.h      |   8 +
 .../MGIS/Behaviour/MaterialStateManager.h     |  28 +++
 bindings/c/src/MaterialDataManager.cxx        |  18 ++
 bindings/c/src/MaterialStateManager.cxx       |  12 ++
 bindings/c/tests/CMakeLists.txt               |  18 ++
 bindings/c/tests/LayoutTest-c.c               | 172 ++++++++++++++++++
 bindings/fortran/src/mgis_behaviour.f90       |  82 +++++++++
 bindings/fortran/tests/CMakeLists.txt         |   1 +
 bindings/fortran/tests/LayoutTest-f.f90       |  62 +++++++
 .../include/MGIS/Python/NumPySupport.hxx      |  20 ++
 bindings/python/src/MaterialDataManager.cxx   |  15 ++
 bindings/python/src/MaterialStateManager.cxx  |  14 +-
 bindings/python/src/NumPySupport.cxx          |  19 ++
 bindings/python/tests/CMakeLists.txt          |   1 +
 bindings/python/tests/LayoutTest.py           |  56 ++++++
 15 files changed, 521 insertions(+), 5 deletions(-)
 create mode 100644 bindings/c/tests/LayoutTest-c.c
 create mode 100644 bindings/fortran/tests/LayoutTest-f.f90
 create mode 100644 bindings/python/tests/LayoutTest.py

diff --git a/bindings/c/include/MGIS/Behaviour/MaterialDataManager.h b/bindings/c/include/MGIS/Behaviour/MaterialDataManager.h
index 59c88df..a26549f 100644
--- a/bindings/c/include/MGIS/Behaviour/MaterialDataManager.h
+++ b/bindings/c/include/MGIS/Behaviour/MaterialDataManager.h
@@ -66,6 +66,14 @@ mgis_status mgis_bv_material_data_manager_initializer_bind_tangent_operator(
 MGIS_C_EXPORT
 mgis_status mgis_bv_material_data_manager_initializer_bind_speed_of_sound(
     mgis_bv_MaterialDataManagerInitializer*, mgis_real* const, mgis_size_type);
+/*!
+ * \brief set the layout of the arrays of the material data manager
+ * \param[in,out] d: initializer
+ * \param[in] l: layout
+ */
+MGIS_C_EXPORT
+mgis_status mgis_bv_material_data_manager_initializer_set_layout(
+    mgis_bv_MaterialDataManagerInitializer*, const mgis_bv_FieldLayout);
 /*!
  * \brief set the state at the beginning of the time step
  * \param[out] s: pointer to a pointer to the state initializer
diff --git a/bindings/c/include/MGIS/Behaviour/MaterialStateManager.h b/bindings/c/include/MGIS/Behaviour/MaterialStateManager.h
index 86e83b8..b4d1569 100644
--- a/bindings/c/include/MGIS/Behaviour/MaterialStateManager.h
+++ b/bindings/c/include/MGIS/Behaviour/MaterialStateManager.h
@@ -52,6 +52,25 @@ typedef enum {
   MGIS_BV_UPDATE_POLICY = 1
 } mgis_bv_MaterialStateManagerUpdatePolicy;
 
+/*!
+ * \brief layout of the arrays of a material state manager (see
+ * `mgis::FieldLayout`)
+ */
+typedef enum {
+  /*!
+   * \brief the components of a point are contiguous: the `j`-th component of
+   * the `i`-th point is stored at `values[i * stride + j]`, where `stride` is
+   * the number of values of a point
+   */
+  MGIS_BV_POINT_BY_POINT = 0,
+  /*!
+   * \brief the values of a component are contiguous: the `j`-th component of
+   * the `i`-th point is stored at `values[j * n + i]`, where `n` is the
+   * number of points
+   */
+  MGIS_BV_COMPONENT_BY_COMPONENT = 1
+} mgis_bv_FieldLayout;
+
 /*!
  * \brief bind the gradients to the given array
  * \param[in,out] s: initializer
@@ -116,9 +135,18 @@ mgis_bv_material_state_manager_get_number_of_integration_points(
  */
 MGIS_C_EXPORT mgis_status mgis_bv_material_state_manager_get_gradients(
     mgis_real**, mgis_bv_MaterialStateManager* const);
+/*!
+ * \param[out] l: a pointer where the layout of the arrays will be written
+ * \param[in]  s: state manager
+ */
+MGIS_C_EXPORT mgis_status mgis_bv_material_state_manager_get_layout(
+    mgis_bv_FieldLayout* const, mgis_bv_MaterialStateManager* const);
 /*!
  * \param[out] gs: a pointer where the gradients' stride will be written
  * \param[in]  s: state manager
+ * \note the stride is the number of values of a point. It is the stride
+ * between two points only if the layout is `MGIS_BV_POINT_BY_POINT` (see
+ * `mgis_bv_material_state_manager_get_layout`).
  */
 MGIS_C_EXPORT mgis_status mgis_bv_material_state_manager_get_gradients_stride(
     mgis_size_type* const, mgis_bv_MaterialStateManager* const);
diff --git a/bindings/c/src/MaterialDataManager.cxx b/bindings/c/src/MaterialDataManager.cxx
index bc861a6..b1852c7 100644
--- a/bindings/c/src/MaterialDataManager.cxx
+++ b/bindings/c/src/MaterialDataManager.cxx
@@ -54,6 +54,24 @@ mgis_status mgis_bv_material_data_manager_initializer_bind_tangent_operator(
   return mgis_report_success();
 }  // end of mgis_bv_material_data_manager_initializer_bind_tangent_operator
 
+mgis_status mgis_bv_material_data_manager_initializer_set_layout(
+    mgis_bv_MaterialDataManagerInitializer* d, const mgis_bv_FieldLayout l) {
+  if (d == nullptr) {
+    return mgis_report_failure(
+        "mgis_bv_material_data_manager_initializer_set_layout: "
+        "null argument");
+  }
+  if ((l != MGIS_BV_POINT_BY_POINT) && (l != MGIS_BV_COMPONENT_BY_COMPONENT)) {
+    return mgis_report_failure(
+        "mgis_bv_material_data_manager_initializer_set_layout: "
+        "invalid layout");
+  }
+  d->layout = l == MGIS_BV_POINT_BY_POINT
+                  ? mgis::FieldLayout::POINT_BY_POINT
+                  : mgis::FieldLayout::COMPONENT_BY_COMPONENT;
+  return mgis_report_success();
+}  // end of mgis_bv_material_data_manager_initializer_set_layout
+
 mgis_status mgis_bv_material_data_manager_initializer_bind_speed_of_sound(
     mgis_bv_MaterialDataManagerInitializer* d,
     mgis_real* const p,
diff --git a/bindings/c/src/MaterialStateManager.cxx b/bindings/c/src/MaterialStateManager.cxx
index e2bade7..a629b2d 100644
--- a/bindings/c/src/MaterialStateManager.cxx
+++ b/bindings/c/src/MaterialStateManager.cxx
@@ -213,6 +213,18 @@ mgis_status mgis_bv_material_state_manager_get_internal_state_variables(
   return mgis_report_success();
 }  // end of mgis_bv_material_state_manager_get_internal_state_variables
 
+mgis_status mgis_bv_material_state_manager_get_layout(
+    mgis_bv_FieldLayout* const l, mgis_bv_MaterialStateManager* const s) {
+  if (s == nullptr) {
+    *l = MGIS_BV_POINT_BY_POINT;
+    return mgis_report_failure("null state manager");
+  }
+  *l = s->layout == mgis::FieldLayout::POINT_BY_POINT
+           ? MGIS_BV_POINT_BY_POINT
+           : MGIS_BV_COMPONENT_BY_COMPONENT;
+  return mgis_report_success();
+}  // end of mgis_bv_material_state_manager_get_layout
+
 mgis_status mgis_bv_material_state_manager_get_internal_state_variables_stride(
     mgis_size_type* const gs, mgis_bv_MaterialStateManager* const s) {
   if (s == nullptr) {
diff --git a/bindings/c/tests/CMakeLists.txt b/bindings/c/tests/CMakeLists.txt
index 46c8b5f..27b7e4b 100644
--- a/bindings/c/tests/CMakeLists.txt
+++ b/bindings/c/tests/CMakeLists.txt
@@ -61,6 +61,11 @@ add_executable(IntegrateTest-c
   IntegrateTest-c.c)
 target_link_libraries(IntegrateTest-c
   PRIVATE MFrontGenericInterface-c MFrontGenericInterface)
+add_executable(LayoutTest-c
+  EXCLUDE_FROM_ALL
+  LayoutTest-c.c)
+target_link_libraries(LayoutTest-c
+  PRIVATE MFrontGenericInterface-c MFrontGenericInterface)
 add_executable(IntegrateTest2-c
   EXCLUDE_FROM_ALL
   IntegrateTest2-c.c)
@@ -248,6 +253,19 @@ else((CMAKE_HOST_WIN32) AND (NOT MSYS))
     PROPERTY DEPENDS BehaviourTest)
 endif((CMAKE_HOST_WIN32) AND (NOT MSYS))
 
+add_test(NAME LayoutTest-c
+ COMMAND LayoutTest-c
+ "$<TARGET_FILE:BehaviourTest>")
+add_dependencies(build-tests LayoutTest-c)
+if((CMAKE_HOST_WIN32) AND (NOT MSYS))
+  set_property(TEST LayoutTest-c
+    PROPERTY DEPENDS BehaviourTest
+    PROPERTY ENVIRONMENT "PATH=$<TARGET_FILE_DIR:MFrontGenericInterface>\;$<TARGET_FILE_DIR:MFrontGenericInterface-c>\;${MGIS_PATH_STRING}")
+else((CMAKE_HOST_WIN32) AND (NOT MSYS))
+  set_property(TEST LayoutTest-c
+    PROPERTY DEPENDS BehaviourTest)
+endif((CMAKE_HOST_WIN32) AND (NOT MSYS))
+
 add_test(NAME IntegrateTest4-c
  COMMAND IntegrateTest4-c
  "$<TARGET_FILE:ModelTest>")
diff --git a/bindings/c/tests/LayoutTest-c.c b/bindings/c/tests/LayoutTest-c.c
new file mode 100644
index 0000000..7b2d842
--- /dev/null
+++ b/bindings/c/tests/LayoutTest-c.c
@@ -0,0 +1,172 @@
+/*!
+ * \file   LayoutTest-c.c
+ * \brief  integration of a behaviour with a material data manager whose
+ * arrays are stored component by component, through the C bindings: the
+ * results are the ones of the test `IntegrateTest3-c`, whose arrays are
+ * stored point by point
+ * \date   08/10/2026
+ * \copyright Use, modification and distribution are subject
+ * to one of the following licences:
+ * - GNU Lesser General Public License (LGPL), Version 3.0. (See accompanying
+ *   file LGPL-3.0.txt)
+ * - CECILL-C,  Version 1.0 (See accompanying files
+ *   CeCILL-C_V1-en.txt and CeCILL-C_V1-fr.txt).
+ */
+
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "MGIS/ThreadPool.h"
+#include "MGIS/Behaviour/MaterialDataManager.h"
+#include "MGIS/Behaviour/Integrate.h"
+
+int test_status = EXIT_SUCCESS;
+mgis_bv_Behaviour* b = NULL;
+mgis_ThreadPool* p = NULL;
+mgis_bv_MaterialDataManager* m = NULL;
+mgis_bv_MaterialDataManagerInitializer* mi = NULL;
+
+static void check_status(const mgis_status s) {
+  if (s.exit_status != MGIS_SUCCESS) {
+    fprintf(stderr, "invalid function call: %s\n", s.msg);
+    mgis_bv_free_behaviour(&b);
+    mgis_bv_free_material_data_manager(&m);
+    mgis_free_thread_pool(&p);
+    exit(EXIT_FAILURE);
+  }
+}  // end of check_status
+
+static int check(const int b, const char* const e) {
+  if (b == 0) {
+    test_status = EXIT_FAILURE;
+    fprintf(stderr, "%s\n", e);
+  }
+  return b;
+}  // end of check
+
+int main(const int argc, const char* const* argv) {
+  mgis_size_type o;
+  if (check(argc == 2, "expected two arguments") == 0) {
+    return EXIT_FAILURE;
+  }
+  const mgis_real p_ref[21] = {0,
+                               1.3523277308229e-11,
+                               1.0955374667213e-07,
+                               5.5890770166084e-06,
+                               3.2392193670428e-05,
+                               6.645865307584e-05,
+                               9.9676622883138e-05,
+                               0.00013302758358953,
+                               0.00016635821069889,
+                               0.00019969195920296,
+                               0.00023302522883648,
+                               0.00026635857194317,
+                               0.000299691903777,
+                               0.0003330252373404,
+                               0.00036635857063843,
+                               0.00039969190397718,
+                               0.00043302523730968,
+                               0.00046635857064314,
+                               0.00049969190397646,
+                               0.00053302523730979,
+                               0.00056635857064313};
+  const mgis_real de = 5.e-5; /* strain increment */
+  const mgis_real dt = 180;
+  mgis_real* g;            /* gradients */
+  mgis_size_type g_stride; /* gradients size */
+  mgis_real*
+      isvs0; /* internal state variables at the beginning of the time step */
+  mgis_real* isvs1; /* internal state variables at the end of the time step */
+  mgis_size_type isvs_stride; /* internal state variables stride */
+  const mgis_size_type n = 100;
+  mgis_bv_MaterialStateManager*
+      s0; /* state at the beginning of the time step */
+  mgis_bv_MaterialStateManager* s1; /* state at the end of the time step */
+  mgis_real pi[21]; /* values of the equivalent plastic strain for the first
+                       integration point */
+  mgis_real pe[21]; /* values of the equivalent plastic strain for the last
+                       integration point */
+  mgis_size_type ni, ne;
+  mgis_size_type idx;
+  mgis_size_type i;
+  int r;
+  mgis_bv_FieldLayout l;
+  check_status(mgis_create_thread_pool(&p, 2));
+  check_status(mgis_bv_load_behaviour(&b, argv[1], "Norton", "Tridimensional"));
+  /* arrays stored component by component */
+  check_status(mgis_bv_create_material_data_manager_initializer(&mi));
+  check_status(mgis_bv_material_data_manager_initializer_set_layout(
+      mi, MGIS_BV_COMPONENT_BY_COMPONENT));
+  check_status(
+      mgis_bv_create_material_data_manager_with_initializer(&m, b, 100, mi));
+  check_status(mgis_bv_behaviour_get_internal_state_variable_offset_by_name(
+      &o, b, "EquivalentViscoplasticStrain"));
+  check_status(mgis_bv_behaviour_get_internal_state_variable_offset_by_name(
+      &o, b, "EquivalentViscoplasticStrain"));
+  check_status(mgis_bv_material_data_manager_get_state_0(&s0, m));
+  check_status(mgis_bv_material_data_manager_get_state_1(&s1, m));
+  check_status(mgis_bv_material_state_manager_get_layout(&l, s1));
+  check(l == MGIS_BV_COMPONENT_BY_COMPONENT, "invalid layout");
+  /* initialize the external state variable */
+  check_status(
+      mgis_bv_material_state_manager_set_uniform_scalar_external_state_variable(
+          s1, "Temperature", 293.15));
+  /* copy s1 in s0 */
+  check_status(mgis_bv_update_material_data_manager(m));
+  check_status(mgis_bv_material_state_manager_get_gradients(&g, s1));
+  check_status(
+      mgis_bv_material_state_manager_get_gradients_stride(&g_stride, s1));
+  check_status(
+      mgis_bv_material_state_manager_get_internal_state_variables(&isvs0, s0));
+  check_status(
+      mgis_bv_material_state_manager_get_internal_state_variables(&isvs1, s1));
+  check_status(
+      mgis_bv_material_state_manager_get_internal_state_variables_stride(
+          &isvs_stride, s1));
+  /* the strides are the number of values of a point */
+  check(g_stride == 6, "invalid number of gradients");
+  check(isvs_stride == 7, "invalid number of internal state variables");
+  /* the first component of the gradients is stored in the first n values */
+  for (idx = 0; idx != n; ++idx) {
+    g[idx] = de;
+  }
+  /* the values of the equivalent plastic strain are stored contiguously */
+  ni = o * n;
+  ne = o * n + n - 1;
+  /* // integration */
+  pi[0] = isvs0[ni];
+  pe[0] = isvs0[ne];
+  for (i = 0; i != 20; ++i) {
+    check_status(mgis_bv_integrate_material_data_manager(
+        &r, p, m, MGIS_BV_INTEGRATION_NO_TANGENT_OPERATOR, dt));
+    check_status(mgis_bv_update_material_data_manager(m));
+    for (idx = 0; idx != n; ++idx) {
+      g[idx] += de;
+    }
+    pi[i + 1] = isvs1[ni];
+    pe[i + 1] = isvs1[ne];
+  }
+  check_status(mgis_bv_free_behaviour(&b));
+  check_status(mgis_bv_free_material_data_manager(&m));
+  check_status(mgis_bv_free_material_data_manager_initializer(&mi));
+  check_status(mgis_free_thread_pool(&p));
+  for (i = 0; i != 21; ++i) {
+    if (fabs(pi[i] - p_ref[i]) > 1.e-12) {
+      fprintf(stderr,
+              "LayoutTest: invalid value for the equivalent "
+              "viscoplastic strain at the first integration point"
+              "(expected '%g', computed '%g')\n",
+              p_ref[i], pi[i]);
+      return EXIT_FAILURE;
+    }
+    if (fabs(pe[i] - p_ref[i]) > 1.e-12) {
+      fprintf(stderr,
+              "LayoutTest: invalid value for the equivalent "
+              "viscoplastic strain at the first integration point"
+              "(expected '%g', computed '%g')\n",
+              p_ref[i], pe[i]);
+      return EXIT_FAILURE;
+    }
+  }
+  return EXIT_SUCCESS;
+}
diff --git a/bindings/fortran/src/mgis_behaviour.f90 b/bindings/fortran/src/mgis_behaviour.f90
index dab70b7..b034788 100644
--- a/bindings/fortran/src/mgis_behaviour.f90
+++ b/bindings/fortran/src/mgis_behaviour.f90
@@ -42,6 +42,11 @@ module mgis_behaviour
      enumerator :: UPDATE = 0
      enumerator :: NOUPDATE = 1
   end enum
+  ! layout of the arrays of a material state manager
+  enum, bind(C)
+     enumerator :: POINT_BY_POINT = 0
+     enumerator :: COMPONENT_BY_COMPONENT = 1
+  end enum
   enum, bind(C)
      enumerator :: PREDICTION_TANGENT_OPERATOR = -3
      enumerator :: PREDICTION_SECANT_OPERATOR = -2
@@ -2737,6 +2742,28 @@ contains
     s = material_data_manager_initializer_bind_speed_of_sound_wrapper(d%ptr,p_ptr,nc)
   end function material_data_manager_initializer_bind_speed_of_sound
   !
+  function material_data_manager_initializer_set_layout(d,l) result(s)
+    use, intrinsic :: iso_c_binding, only: c_int
+    use mgis, only: mgis_status
+    implicit none
+    interface
+       function material_data_manager_initializer_set_layout_wrapper(d,l) &
+            bind(c,name = 'mgis_bv_material_data_manager_initializer_set_layout') &
+            result(r)
+         use, intrinsic :: iso_c_binding, only: c_ptr, c_int
+         use mgis, only: mgis_status
+         implicit none
+         type(c_ptr), intent(in), value :: d
+         integer(kind=c_int), intent(in), value :: l
+         type(mgis_status) :: r
+       end function material_data_manager_initializer_set_layout_wrapper
+    end interface
+    type(MaterialDataManagerInitializer), intent(in) :: d
+    integer(kind=c_int), intent(in) :: l
+    type(mgis_status) :: s
+    s = material_data_manager_initializer_set_layout_wrapper(d%ptr,l)
+  end function material_data_manager_initializer_set_layout
+  !
   function free_material_data_manager_initializer(ptr) result(r)
     use, intrinsic :: iso_c_binding, only: c_associated
     use mgis
@@ -3100,6 +3127,49 @@ contains
     s = material_data_manager_get_state_1_wrapper(s1%ptr, d%ptr)
   end function material_data_manager_get_state_1
   !
+  function material_state_manager_get_layout(l, s) &
+       result(r)
+    use, intrinsic :: iso_c_binding, only: c_int
+    use mgis, only: mgis_status
+    implicit none
+    interface
+       function material_state_manager_get_layout_wrapper(l, s) &
+            bind(c,name = 'mgis_bv_material_state_manager_get_layout') &
+            result(r)
+         use, intrinsic :: iso_c_binding, only: c_ptr, c_int
+         use mgis, only: mgis_status
+         integer(kind=c_int), intent(out) :: l
+         type(c_ptr), intent(in),value :: s
+         type(mgis_status) :: r
+       end function material_state_manager_get_layout_wrapper
+    end interface
+    integer(kind=c_int), intent(out) :: l
+    type(MaterialStateManager), intent(in) :: s
+    type(mgis_status) :: r
+    r = material_state_manager_get_layout_wrapper(l, s %ptr)
+  end function material_state_manager_get_layout
+  !
+  ! check that the arrays of a material state manager are stored point by
+  ! point, as required by the functions returning them as two-dimensional
+  ! arrays (one column per point)
+  function material_state_manager_check_point_by_point_layout(s, f) &
+       result(r)
+    use, intrinsic :: iso_c_binding, only: c_int
+    use mgis, only: mgis_status, report_failure, MGIS_SUCCESS
+    implicit none
+    type(MaterialStateManager), intent(in) :: s
+    character(len=*), intent(in) :: f
+    type(mgis_status) :: r
+    integer(kind=c_int) l
+    r = material_state_manager_get_layout(l, s)
+    if( r%exit_status .ne. MGIS_SUCCESS) then
+       return
+    end if
+    if (l .ne. POINT_BY_POINT) then
+       r = report_failure(f // ": the arrays are stored component by component")
+    end if
+  end function material_state_manager_check_point_by_point_layout
+  !
   function material_state_manager_get_number_of_integration_points(n, s) &
        result(r)
     use, intrinsic :: iso_c_binding, only: c_size_t
@@ -3194,6 +3264,10 @@ contains
     if( r%exit_status .ne. MGIS_SUCCESS) then
        return
     end if
+    r = material_state_manager_check_point_by_point_layout(s, "material_state_manager_get_gradients")
+    if( r%exit_status .ne. MGIS_SUCCESS) then
+       return
+    end if
     call c_f_pointer(g_ptr, g, [gs, n])
   end function material_state_manager_get_gradients
   !
@@ -3262,6 +3336,10 @@ contains
     if( r%exit_status .ne. MGIS_SUCCESS) then
        return
     end if
+    r = material_state_manager_check_point_by_point_layout(s, "material_state_manager_get_thermodynamic_forces")
+    if( r%exit_status .ne. MGIS_SUCCESS) then
+       return
+    end if
     call c_f_pointer(tf_ptr, tf, [tfs, n])
   end function material_state_manager_get_thermodynamic_forces
   !
@@ -3501,6 +3579,10 @@ contains
     if( r%exit_status .ne. MGIS_SUCCESS) then
        return
     end if
+    r = material_state_manager_check_point_by_point_layout(s, "material_state_manager_get_internal_state_variables")
+    if( r%exit_status .ne. MGIS_SUCCESS) then
+       return
+    end if
     call c_f_pointer(isvs_ptr, isvs, [n_isvs, n])
   end function material_state_manager_get_internal_state_variables
   !
diff --git a/bindings/fortran/tests/CMakeLists.txt b/bindings/fortran/tests/CMakeLists.txt
index c5990cc..4d90116 100644
--- a/bindings/fortran/tests/CMakeLists.txt
+++ b/bindings/fortran/tests/CMakeLists.txt
@@ -59,6 +59,7 @@ test_fortran_bindings(StateTest2-f)
 test_fortran_bindings(IntegrateTest-f)
 test_fortran_bindings(IntegrateTest2-f)
 test_fortran_bindings(IntegrateTest3-f)
+test_fortran_bindings(LayoutTest-f)
 test_fortran_bindings(IntegrateTest4-f)
 test_fortran_bindings(IntegrateTest5-f)
 test_fortran_bindings(ParameterTest-f)
diff --git a/bindings/fortran/tests/LayoutTest-f.f90 b/bindings/fortran/tests/LayoutTest-f.f90
new file mode 100644
index 0000000..d4dff59
--- /dev/null
+++ b/bindings/fortran/tests/LayoutTest-f.f90
@@ -0,0 +1,62 @@
+!
+! \file   LayoutTest-f.f90
+! \brief  material data manager whose arrays are stored component by
+!         component, through the fortran bindings: the layout is set by the
+!         initializer, the behaviour is integrated, and the functions
+!         returning the arrays as two-dimensional arrays (one column per
+!         point) report an error
+! \date   08/10/2026
+!
+subroutine test()
+  use, intrinsic :: iso_c_binding, only: c_int
+  use mgis
+  use mgis_behaviour
+  use mgis_testing_utilities
+  implicit none
+  type(ThreadPool) p
+  type(Behaviour) b
+  type(MaterialDataManagerInitializer) i
+  type(MaterialDataManager) m
+  type(MaterialStateManager) s1
+  type(mgis_status) st
+  real(kind=8), dimension (:,:), pointer :: g => null()
+  integer(kind=c_int) :: l
+  integer :: n = 10
+  integer :: ri
+  logical :: r
+  call check_status(create_thread_pool(p, 2))
+  call check_status(load_behaviour(b, &
+       get_mfront_behaviour_test_library_path(), &
+       'Norton', 'Tridimensional'))
+  call check_status(create_material_data_manager_initializer(i))
+  call check_status(material_data_manager_initializer_set_layout(i, &
+       COMPONENT_BY_COMPONENT))
+  call check_status(create_material_data_manager_with_initializer(m, b, i, n))
+  call check_status(material_data_manager_get_state_1(s1, m))
+  call check_status(material_state_manager_get_layout(l, s1))
+  r = check(l == COMPONENT_BY_COMPONENT, 'invalid layout')
+  call check_status(material_state_manager_set_us_external_state_variable( &
+       s1, "Temperature", 293.15d0))
+  call check_status(update_material_data_manager(m))
+  call check_status(integrate_material_data_manager( &
+       ri, p, m, INTEGRATION_NO_TANGENT_OPERATOR, 180d0))
+  r = check(ri.eq.1, 'integration failed')
+  ! the arrays can't be returned as two-dimensional arrays
+  st = material_state_manager_get_gradients(g, s1)
+  r = check(st%exit_status .ne. MGIS_SUCCESS, &
+       'the layout of the gradients has not been checked')
+  ! free ressources
+  call check_status(free_material_data_manager(m))
+  call check_status(free_material_data_manager_initializer(i))
+  call check_status(free_behaviour(b))
+  call check_status(free_thread_pool(p))
+end subroutine test
+
+program main
+  use mgis_testing_utilities
+  call test()
+  call tests_summary()
+  if(.not. status) then
+     stop -1
+  end if
+end program main
diff --git a/bindings/python/include/MGIS/Python/NumPySupport.hxx b/bindings/python/include/MGIS/Python/NumPySupport.hxx
index 7cd7b48..42d9820 100644
--- a/bindings/python/include/MGIS/Python/NumPySupport.hxx
+++ b/bindings/python/include/MGIS/Python/NumPySupport.hxx
@@ -21,6 +21,7 @@
 #include <pybind11/pybind11.h>
 #include <pybind11/numpy.h>
 #include "MGIS/Config.hxx"
+#include "MGIS/FieldView.hxx"
 
 namespace mgis::python {
 
@@ -111,6 +112,25 @@ namespace mgis::python {
 
   std::span<mgis::real> mgis_convert_to_span(const pybind11::array_t<double>&);
 
+  /*!
+   * \brief create a 2D-ndarray object (points, components) from a view,
+   * with the strides of the view (point by point or component by
+   * component). The ndarray does not own the data.
+   * \param[in] v: view
+   */
+  pybind11::array_t<double> wrapInNumPyArray(const FieldView<double>&);
+  /*!
+   * \brief create a 3D-ndarray object (points, rows, columns) from a view
+   * whose components are the values of a matrix stored row by row, with the
+   * strides of the view. The ndarray does not own the data.
+   * \param[in] v: view
+   * \param[in] nr: number of rows
+   * \param[in] nc: number of columns
+   */
+  pybind11::array_t<double> wrapInNumPyArray(const FieldView<double>&,
+                                             const mgis::size_type,
+                                             const mgis::size_type);
+
 }  // end of namespace mgis::python
 
 #endif /* LIB_MGIS_PYTHON_NUMPYSUPPORT_HXX */
diff --git a/bindings/python/src/MaterialDataManager.cxx b/bindings/python/src/MaterialDataManager.cxx
index 95226d9..18486b2 100644
--- a/bindings/python/src/MaterialDataManager.cxx
+++ b/bindings/python/src/MaterialDataManager.cxx
@@ -43,6 +43,18 @@ static void MaterialDataManager_useExternalArrayOfSpeedOfSounds(
 
 static pybind11::object MaterialDataManager_getK(
     mgis::behaviour::MaterialDataManager& d) {
+  if (!d.K.empty()) {
+    // the tangent operator blocks, in the layout of the manager
+    const auto K = getTangentOperatorBlocksView(d);
+    if (d.b.to_blocks.size() == 1u) {
+      const auto nl =
+          getVariableSize(d.b.to_blocks.front().second, d.b.hypothesis);
+      const auto nc =
+          getVariableSize(d.b.to_blocks.front().first, d.b.hypothesis);
+      return mgis::python::wrapInNumPyArray(K, nc, nl);
+    }
+    return mgis::python::wrapInNumPyArray(K);
+  }
   if (d.b.to_blocks.size() == 1u) {
     const auto nl =
         getVariableSize(d.b.to_blocks.front().second, d.b.hypothesis);
@@ -65,8 +77,10 @@ void declareMaterialDataManager(pybind11::module_& m) {
   // exporting the MaterialDataManager class
   pybind11::class_<MaterialDataManagerInitializer>(
       m, "MaterialDataManagerInitializer")
+      .def(pybind11::init<>())
       .def_readonly("s0", &MaterialDataManagerInitializer::s0)
       .def_readonly("s1", &MaterialDataManagerInitializer::s1)
+      .def_readwrite("layout", &MaterialDataManagerInitializer::layout)
       .def("bindTangentOperator",
            &MaterialDataManagerInitializer_bindTangentOperator,
            "use the given array to store the tangent operator blocks")
@@ -101,6 +115,7 @@ void declareMaterialDataManager(pybind11::module_& m) {
            "release the array of speed of sounds")
       .def_readonly("n", &MaterialDataManager::n)
       .def_readonly("number_of_integration_points", &MaterialDataManager::n)
+      .def_readonly("layout", &MaterialDataManager::layout)
       .def_readonly("s0", &MaterialDataManager::s0)
       .def_readonly("s1", &MaterialDataManager::s1)
       .def_property_readonly("K", &MaterialDataManager_getK)
diff --git a/bindings/python/src/MaterialStateManager.cxx b/bindings/python/src/MaterialStateManager.cxx
index 0d5c16f..6e3c244 100644
--- a/bindings/python/src/MaterialStateManager.cxx
+++ b/bindings/python/src/MaterialStateManager.cxx
@@ -46,19 +46,17 @@ static void MaterialStateManagerInitializer_bindDissipatedEnergies(
 
 static pybind11::object MaterialStateManager_getGradients(
     mgis::behaviour::MaterialStateManager& s) {
-  return mgis::python::wrapInNumPyArray(s.gradients, s.gradients_stride);
+  return mgis::python::wrapInNumPyArray(getGradientsView(s));
 }  // end of MaterialStateManager_getGradients
 
 static pybind11::object MaterialStateManager_getThermodynamicForces(
     mgis::behaviour::MaterialStateManager& s) {
-  return mgis::python::wrapInNumPyArray(s.thermodynamic_forces,
-                                        s.thermodynamic_forces_stride);
+  return mgis::python::wrapInNumPyArray(getThermodynamicForcesView(s));
 }  // end of MaterialStateManager_getThermodynamicForces
 
 static pybind11::object MaterialStateManager_getInternalStateVariables(
     mgis::behaviour::MaterialStateManager& s) {
-  return mgis::python::wrapInNumPyArray(s.internal_state_variables,
-                                        s.internal_state_variables_stride);
+  return mgis::python::wrapInNumPyArray(getInternalStateVariablesView(s));
 }  // end of MaterialStateManager_getInternalStateVariables
 
 static pybind11::object MaterialStateManager_getStoredEnergies(
@@ -214,6 +212,11 @@ void declareMaterialStateManager(pybind11::module_& m) {
       .def("bindDissipatedEnergies",
            &MaterialStateManagerInitializer_bindDissipatedEnergies,
            "use the given array to store the dissipated energies");
+  // layout of the arrays
+  pybind11::enum_<mgis::FieldLayout>(m, "FieldLayout")
+      .value("POINT_BY_POINT", mgis::FieldLayout::POINT_BY_POINT)
+      .value("COMPONENT_BY_COMPONENT",
+             mgis::FieldLayout::COMPONENT_BY_COMPONENT);
   // wrapping the MaterialStateManager class
   pybind11::class_<MaterialStateManager>(m, "MaterialStateManager")
       .def(pybind11::init<const Behaviour&, const size_type>())
@@ -221,6 +224,7 @@ void declareMaterialStateManager(pybind11::module_& m) {
                           const MaterialStateManagerInitializer&>())
       .def_readonly("n", &MaterialStateManager::n)
       .def_readonly("number_of_integration_points", &MaterialStateManager::n)
+      .def_readonly("layout", &MaterialStateManager::layout)
       .def_property_readonly("gradients", &MaterialStateManager_getGradients)
       .def_readonly("gradients_stride", &MaterialStateManager::gradients_stride)
       .def_property_readonly("thermodynamic_forces",
diff --git a/bindings/python/src/NumPySupport.cxx b/bindings/python/src/NumPySupport.cxx
index 35b9f6f..947509a 100644
--- a/bindings/python/src/NumPySupport.cxx
+++ b/bindings/python/src/NumPySupport.cxx
@@ -93,4 +93,23 @@ namespace mgis::python {
     return {static_cast<double*>(i.ptr), static_cast<mgis_size_type>(i.size)};
   }  // end of mgis_convert_to_span
 
+  pybind11::array_t<double> wrapInNumPyArray(const FieldView<double>& v) {
+    constexpr auto s = sizeof(double);
+    return pybind11::array_t<double>(
+        {v.getNumberOfPoints(), v.getNumberOfComponents()},
+        {v.getPointStride() * s, v.getComponentStride() * s}, v.data(),
+        pybind11::none());
+  }  // end of wrapInNumPyArray
+
+  pybind11::array_t<double> wrapInNumPyArray(const FieldView<double>& v,
+                                             const size_type nr,
+                                             const size_type nc) {
+    constexpr auto s = sizeof(double);
+    return pybind11::array_t<double>(
+        {v.getNumberOfPoints(), nr, nc},
+        {v.getPointStride() * s, nc * v.getComponentStride() * s,
+         v.getComponentStride() * s},
+        v.data(), pybind11::none());
+  }  // end of wrapInNumPyArray
+
 }  // end of namespace mgis::python
diff --git a/bindings/python/tests/CMakeLists.txt b/bindings/python/tests/CMakeLists.txt
index 8b7933d..9047150 100644
--- a/bindings/python/tests/CMakeLists.txt
+++ b/bindings/python/tests/CMakeLists.txt
@@ -37,6 +37,7 @@ test_python_bindings(ParameterTest)
 test_python_bindings(IntegrateTest)
 test_python_bindings(IntegrateTest2)
 test_python_bindings(IntegrateTest3)
+test_python_bindings(LayoutTest)
 test_python_bindings(IntegrateTest4)
 test_python_bindings(IntegrateTest5)
 test_python_bindings(IntegrateTest6)
diff --git a/bindings/python/tests/LayoutTest.py b/bindings/python/tests/LayoutTest.py
new file mode 100644
index 0000000..8a118c1
--- /dev/null
+++ b/bindings/python/tests/LayoutTest.py
@@ -0,0 +1,56 @@
+# \file   LayoutTest.py
+# \brief  material data managers whose arrays are stored point by point and
+#         component by component: same results, numpy arrays with the
+#         strides of the layout
+# \date   08/10/2026
+
+import os
+try:
+    import unittest2 as unittest
+except ImportError:
+    import unittest
+import mgis
+import mgis.behaviour as mgis_bv
+
+
+class LayoutTest(unittest.TestCase):
+    def test_pass(self):
+        lib = os.environ['MGIS_TEST_BEHAVIOURS_LIBRARY']
+        h = mgis_bv.Hypothesis.Tridimensional
+        b = mgis_bv.load(lib, 'Norton', h)
+        nig = 10
+        managers = []
+        for layout in [mgis_bv.FieldLayout.POINT_BY_POINT,
+                       mgis_bv.FieldLayout.COMPONENT_BY_COMPONENT]:
+            i = mgis_bv.MaterialDataManagerInitializer()
+            i.layout = layout
+            m = mgis_bv.MaterialDataManager(b, nig, i)
+            self.assertTrue(m.layout == layout)
+            self.assertTrue(m.s1.layout == layout)
+            mgis_bv.setExternalStateVariable(m.s1, 'Temperature', 293.15)
+            mgis_bv.update(m)
+            managers.append(m)
+        # the arrays are seen with the same indices (point, component)
+        g = managers[1].s1.gradients
+        self.assertTrue(g.shape == (nig, 6))
+        self.assertTrue(g.strides == (8, 8 * nig))
+        isvs = managers[1].s1.internal_state_variables
+        self.assertTrue(isvs.strides == (8, 8 * nig))
+        it = mgis_bv.IntegrationType.IntegrationWithConsistentTangentOperator
+        for step in range(0, 3):
+            for m in managers:
+                for j in range(0, nig):
+                    m.s1.gradients[j][0] += 5.e-5 * (1 + 0.1 * j)
+                mgis_bv.integrate(m, it, 180, 0, nig)
+            for a in ['thermodynamic_forces', 'internal_state_variables']:
+                v0 = getattr(managers[0].s1, a)
+                v1 = getattr(managers[1].s1, a)
+                self.assertTrue((v0 == v1).all())
+            self.assertTrue((managers[0].K == managers[1].K).all())
+            for m in managers:
+                mgis_bv.update(m)
+        pass
+
+
+if __name__ == '__main__':
+    unittest.main()
-- 
2.43.0

